### PR TITLE
planner,executor: support `admin show slow` command

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

After this PR, together with https://github.com/pingcap/tidb/pull/7688 and https://github.com/pingcap/tidb/pull/7692, we can use `admin show slow` now.

### What is changed and how it works?

Build `ShowSlow` plan and `ShowSlowExec` executor.

It works like this:

```
mysql> admin show slow top 3;
+--------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------+-----------+----------+
| SQL                            | START                      | DURATION | DETAILS                                                                                              | SUCC | CONN_ID | TRANSACTION_TS     | USER           | DB   | TABLE_IDS | INDEX_IDS | INTERNAL |
+--------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------+-----------+----------+
| select count(*) from t limit 5 | 2018-09-26 16:09:21.101256 | 00:00:09 | process_time:1m8.032s wait_time:31.62s request_count:14 total_keys:20000014 processed_keys:20000000  |    1 |       2 | 403164197311217665 | root@127.0.0.1 | test | [32]      |           |        1 |
| select count(*) from t         | 2018-09-26 16:09:11.269202 | 00:00:08 | process_time:1m2.451s wait_time:29.444s request_count:14 total_keys:20000014 processed_keys:20000000 |    1 |       1 | 403164194729099265 | root@127.0.0.1 | test | [32]      |           |        1 |
+--------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------+-----------+----------+
2 rows in set (0.01 sec)
```

### Check List 

Tests 

 - Unit test to cover the basic statement.

As the result fields always change, it's hard to put it into unit test.
I manually check the result to assure that works.

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

**Comment by winkyao** [comment]
Please fix ci.

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -746,6 +750,23 @@ func buildShowDDLJobQueriesFields() *expression.Schema {
 	return schema
 }
 
+func buildShowSlowSchema() *expression.Schema {
+	schema := expression.NewSchema(make([]*expression.Column, 0, 11)...)
+	schema.Append(buildColumn("", "SQL", mysql.TypeVarchar, 256))
```
Comment (coocood) [inline planner/core/planbuilder.go]: length is too small.

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -746,6 +750,23 @@ func buildShowDDLJobQueriesFields() *expression.Schema {
...
+	schema.Append(buildColumn("", "START", mysql.TypeTimestamp, 64))
+	schema.Append(buildColumn("", "DURATION", mysql.TypeDuration, 64))
+	schema.Append(buildColumn("", "DETAILS", mysql.TypeVarchar, 256))
+	schema.Append(buildColumn("", "SUCC", mysql.TypeTiny, 2))
+	schema.Append(buildColumn("", "CONN_ID", mysql.TypeLonglong, 4))
+	schema.Append(buildColumn("", "TRANSACTION_TS", mysql.TypeLonglong, 4))
+	schema.Append(buildColumn("", "USER", mysql.TypeVarchar, 32))
+	schema.Append(buildColumn("", "DB", mysql.TypeVarchar, 64))
+	schema.Append(buildColumn("", "TABLE_IDS", mysql.TypeVarchar, 64))
```
Comment (coocood) [inline planner/core/planbuilder.go]: length is too small.

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -746,6 +750,23 @@ func buildShowDDLJobQueriesFields() *expression.Schema {
...
+	schema.Append(buildColumn("", "INDEX_IDS", mysql.TypeVarchar, 64))
```
Comment (coocood) [inline planner/core/planbuilder.go]: length is too small.

File: planner/core/planbuilder.go [hunk 4]
```diff
@@ -746,6 +750,23 @@ func buildShowDDLJobQueriesFields() *expression.Schema {
...
 
+func buildShowSlowSchema() *expression.Schema {
+	schema := expression.NewSchema(make([]*expression.Column, 0, 11)...)
+	schema.Append(buildColumn("", "SQL", mysql.TypeVarchar, 256))
+	schema.Append(buildColumn("", "START", mysql.TypeTimestamp, 64))
+	schema.Append(buildColumn("", "DURATION", mysql.TypeDuration, 64))
+	schema.Append(buildColumn("", "DETAILS", mysql.TypeVarchar, 256))
+	schema.Append(buildColumn("", "SUCC", mysql.TypeTiny, 2))
+	schema.Append(buildColumn("", "CONN_ID", mysql.TypeLonglong, 4))
+	schema.Append(buildColumn("", "TRANSACTION_TS", mysql.TypeLonglong, 4))
```
Comment (coocood) [inline planner/core/planbuilder.go]: length is too small

File: planner/core/planbuilder.go [hunk 5]
```diff
@@ -746,6 +750,23 @@ func buildShowDDLJobQueriesFields() *expression.Schema {
...
 }
 
+func buildShowSlowSchema() *expression.Schema {
+	schema := expression.NewSchema(make([]*expression.Column, 0, 11)...)
+	schema.Append(buildColumn("", "SQL", mysql.TypeVarchar, 256))
+	schema.Append(buildColumn("", "START", mysql.TypeTimestamp, 64))
+	schema.Append(buildColumn("", "DURATION", mysql.TypeDuration, 64))
+	schema.Append(buildColumn("", "DETAILS", mysql.TypeVarchar, 256))
+	schema.Append(buildColumn("", "SUCC", mysql.TypeTiny, 2))
+	schema.Append(buildColumn("", "CONN_ID", mysql.TypeLonglong, 4))
```
Comment (coocood) [inline planner/core/planbuilder.go]: length is too small.

File: planner/core/planbuilder.go [hunk 6]
```diff
@@ -746,6 +750,23 @@ func buildShowDDLJobQueriesFields() *expression.Schema {
 	return schema
 }
 
+func buildShowSlowSchema() *expression.Schema {
+	schema := expression.NewSchema(make([]*expression.Column, 0, 11)...)
+	schema.Append(buildColumn("", "SQL", mysql.TypeVarchar, 256))
+	schema.Append(buildColumn("", "START", mysql.TypeTimestamp, 64))
```
Comment (coocood) [inline planner/core/planbuilder.go]: length is too big

File: planner/core/planbuilder.go [hunk 7]
```diff
@@ -746,6 +750,23 @@ func buildShowDDLJobQueriesFields() *expression.Schema {
...
+	schema.Append(buildColumn("", "DURATION", mysql.TypeDuration, 64))
```
Comment (coocood) [inline planner/core/planbuilder.go]: length is too big

File: executor/executor.go [hunk 8]
```diff
@@ -472,6 +474,64 @@ func (e *CheckIndexExec) Next(ctx context.Context, chk *chunk.Chunk) error {
...
+// Next implements the Executor Next interface.
+func (e *ShowSlowExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	if e.done {
+		return nil
+	}
+
+	dom := domain.GetDomain(e.ctx)
+	slowQueries := dom.ShowSlowQuery(e.ShowSlow)
+	for _, slow := range slowQueries {
```
Comment (zz-jason) [inline executor/executor.go]: the number of rows appended to `chk` should be smaller than `max_chunk_size`

File: executor/adapter.go [hunk 9]
```diff
@@ -381,6 +381,12 @@ func (a *ExecStmt) logSlowQuery(txnTS uint64, succ bool) {
 		if user != nil {
 			userString = user.String()
 		}
+		if len(tableIDs) > 10 {
```
Comment (winkyao) [inline executor/adapter.go]: what about `if len(tableIDs) > len("table_ids:")` ?

File: executor/executor.go (hunk 8 above)
Comment (tiancaiamao) [reply executor/executor.go]: So if the result is larger than `max_chunk_size`, it has to be split into multiple batches?

File: executor/show_test.go [hunk 10]
```diff
@@ -637,11 +637,34 @@ func (s *testSuite) TestShowTableStatus(c *C) {
...
+
+func (s *testSuite) TestShowSlow(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec(`drop table if exists t`)
+	tk.MustExec(`create table t(a bigint)`)
+	tk.MustQuery(`select sleep(1)`)
+
+	// Collecting slow queries is asynchronous, wait a while to ensure it's done.
+	time.Sleep(5 * time.Millisecond)
```
Comment (winkyao) [inline executor/show_test.go]: Is 5ms enough?

Comment (coocood) [reply executor/show_test.go]: Maybe add a retry to make the test stable?

File: planner/core/planbuilder.go (hunk 1 above)
Comment (tiancaiamao) [reply planner/core/planbuilder.go]: How about 1024?

File: executor/show_test.go [hunk 11]
```diff
@@ -637,11 +637,34 @@ func (s *testSuite) TestShowTableStatus(c *C) {
...
 	c.Assert(errors.ErrorStack(err), Equals, "")
 	c.Assert(rows[0].GetString(16), Equals, "partitioned")
 }
+
+func (s *testSuite) TestShowSlow(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec(`drop table if exists t`)
+	tk.MustExec(`create table t(a bigint)`)
+	tk.MustQuery(`select sleep(1)`)
```
Comment (winkyao) [inline executor/show_test.go]: We can't assume sleep 1s is a slow query, we need to use cfg.Log.SlowThreshold?

File: executor/executor.go (hunk 8 above)
Comment (zz-jason) [reply executor/executor.go]: Yes.

File: planner/core/planbuilder.go (hunk 6 above)
Comment (tiancaiamao) [reply planner/core/planbuilder.go]: 32?

File: planner/core/planbuilder.go (hunk 1 above)
Comment (winkyao) [reply planner/core/planbuilder.go]: 1024 still too small.

File: executor/executor.go (hunk 8 above)
Comment (zz-jason) [reply executor/executor.go]: we have to use a cursor to record how many slow queries have been returned to the client.

File: executor/adapter.go (hunk 9 above)
Comment (tiancaiamao) [reply executor/adapter.go]: There is a comment for it already.

**Comment by morgo** [comment]
I just started playing with this today :-)

@tiancaiamao Is it possible to change the duration to be in `ms` (unformatted)?  1 second is quite a long time for OLTP, and a query could still be considered slow but show a `00:00:00` duration.

**Pull request merged**